### PR TITLE
[feature] [comment] [contract] comment feature in contract

### LINKED ISSUE #110: [contract][Comment] Delete/Edit Comment: to prove 2 epochKey from the same user.
## Description
After discussing with comment team members, the deleting/edit function for the comment (or even the post) **should emit a delete event to update the comment/post content**. 

## Scenario 1: 
user comment/post at epoch `100` and s/he wants to edit/delete their epoch key in the same epoch. It's relatively simple. We could verify this key with the same epochKeyProof.

## Scenario 2:
user comment/post at epoch `100`, yet s/he wants to edit/delete your content in epoch `107`. Here comes with a problem: **the epochKey would be updated and we could not regenerate the same epochKey. Therefore, it's hard to verify the user identity while editing and deleting the content.**

## Solution
One of the possible solutions of this is to generate a proof that prove 2 epochKey (from different epoch) is generated from the same identity secret without actually recording the relationship among the epochKeys and identity secret.

## Alternatives
Other solutions:
1. limiting the users only update their content in the same epoch period (In my opinion, this is not reasonable for a social media).
2. Storing all past epochKeyProofs.

And this issue might be me overthinking, if there are other solutions/ you do not think this is a issue, please discuss with me. (cc @vivianjeng & @kittybest )

### LINKED ISSUE #126: [contract][Comment] comment storage
**Is your feature request related to a problem? Please describe.**
comment storage in IPFS and backend

**Describe the solution you'd like**
store event output of comment CID in relayer DB，comment + CID also stored in IPFS 

* CID onchain, content in IPFS
* store event output in backend
* event with a content which is CID (same as post)
* delete event ，a potential solution is to emit add comment event and emit delete comment event. wait for #110 solution

**Describe alternatives you've considered**
N/A

**Additional context**
reference: post https://github.com/Unirep/Unirep-Social/blob/2e8e01509a7c6a1aa7f4657666ac006291734eea/packages/backend/src/routes/PostRouter.ts#L374

### LINKED ISSUE #139: [contract] attestedId should be `immutable`
**Describe the bug**
The variable `attesterId` is now muttable, yet it should be act as a constant once assigned in the constructor.

**Solution**
just convert this variable to immutable.

### LINKED ISSUE #164: [contract][comment] bug issue after updating contract
**Describe the bug**
Relay test in branch `feature-contract-comment` now encounters errors in both login and post test.

**To Reproduce**

1. Switch to branch `feature-contract-comment`
2. run `yarn install` && `yarn build`
3. run `yarn relay test`
4. See errors

**Screenshots**
5 errors happen in the test:

```sh
$ yarn workspace @unirep-app/relay run test
$ hardhat test


  LOGIN /login
-----------------------------------------------------------------
Epoch tree depth: 17
State tree depth: 17
History tree depth: 17
Number of epoch keys per epoch: 3
Total fields per user: 6
Sum fields per user: 4
Replacement field nonce bits: 48
Replacement field data bits: 205
-----------------------------------------------------------------
Make sure these match what you expect!
-----------------------------------------------------------------
Deploying EpochKeyVerifier
Deploying ProveReputationVerifier
Deploying UserStateTransitionVerifier
Deploying SignupVerifier
Deploying EpochKeyLiteVerifier
Deploying PreventDoubleActionVerifier
Deploying Unirep
-----------------------------------------------------------------
Bytecode size of Unirep: 16184 bytes
Gas cost of deploying Unirep: 3385237
Deployed to: 0x4D137bb44553d55AE6B28B5391c6f537b06C9cc3
-----------------------------------------------------------------
Deploying EpochKeyVerifier
Deploying EpochKeyVerifierHelper
Deploying EpochKeyLiteVerifier
Deploying EpochKeyLiteVerifierHelper
Unirep app with epoch length 100000 is deployed to 0x959922bE3CAee4b8Cd9a407cc3ac1C251C2007B1
Starting synchronizer...
Synchronizer started
Starting Helia ipfs node...
Helia ipfs node started
Starting transaction manager...
Transaction manager started
Listening on port 8000
    ✔ /api/login, return url
    ✔ /api/user, init user with wrong code and return error
    ✔ /api/user, init user
    ✔ /api/signup, user sign up with wrong proof and return error (473ms)
    ✔ /api/signup, user sign up with wallet (8736ms)
    1) /api/signup, sign up with the same commitment
    ✔ /api/signup, sign up with different attesterId (3781ms)
    ✔ /api/signup, user sign up with server (11866ms)
    2) /api/signup, handle duplicate signup
    3) /api/login, registered user with own wallet
    4) /api/login, registered user with server wallet

  POST /post
    5) "before each" hook for "should create a post"


  7 passing (52s)
  5 failing
1) LOGIN /login
       /api/signup, sign up with the same commitment:

      AssertionError: expected promise to be rejected with 'Error' but it was fulfilled with ''
      + expected - actual

      +Error
      
  

  2) LOGIN /login
       /api/signup, handle duplicate signup:

      AssertionError: expected Response{ domain: null, …(36), …(1) } to have status code 400 but got 200
      + expected - actual

      -200
      +400
      
      at /Users/foodchain/Documents/PSE/social-tw-website/packages/relay/test/login.test.ts:357:37
      at processTicksAndRejections (node:internal/process/task_queues:95:5)
      at async Context.<anonymous> (test/login.test.ts:345:9)

  3) LOGIN /login
       /api/login, registered user with own wallet:

      AssertionError: expected 1 to equal 2
      + expected - actual

      -1
      +2
      
      at Context.<anonymous> (test/login.test.ts:369:42)

  4) LOGIN /login
       /api/login, registered user with server wallet:

      AssertionError: expected 1 to equal 3
      + expected - actual

      -1
      +3
      
      at Context.<anonymous> (test/login.test.ts:378:42)

  5) POST /post
       "before each" hook for "should create a post":

      AssertionError: expected false to equal true
      + expected - actual

      -false
      +true
      
      at Context.<anonymous> (test/post.test.ts:67:29)
```
I reckon this might cause due to the change of Error in smart contract, which were changed by me ( from `require...` to `Customized Errors`). However, since I am not so familiar with the structure in relayer and some of the errors seem like having no relation with the changes in contract, I might need someone who knows the structure of relayer to deal with this.

### PR DESCRIPTION
**Pull request by FoodChain1028**
## Summary

This PR mainly focus on the comment feature in our smart contract. There would be 2 main functions: `leaveComment` and `editComment`.

## Linked Issue

close #110 close #126 close #139 close #164 

## Details

There are two main function in comment feature:
1. `leaveComment`: 
    * should verify a `epochKey` from the server. 
    * should emit a Event 
    ```solidity
    Comment(
        uint256 indexed epochKey,
        uint256 indexed postId,
        uint256 indexed commentId,
        uint256 epoch,
        string content
    );
    ```
2. `editComment`:  
    * should verify a `epochLiteKey` to check the identity
    * should have a input `isRemove` to emit different Event
    * **EDIT**: should emit an event
    ```solidity
    UpdatedComment(
        uint256 indexed epochKey,
        uint256 indexed postId,
        uint256 indexed commentId,
        uint256 epoch,
        string newContent
    );
    ```
    * **REMOVE**: should emit an event 
    ```solidity
    UpdatedComment(
        uint256 indexed epochKey,
        uint256 indexed postId,
        uint256 indexed commentId,
        uint256 epoch,
        string newContent
    );
    ```
    with newContent empty ""

The details of EpochKeyLiteProof can be found [here](https://developer.unirep.io/docs/circuits-api/classes/src.EpochKeyLiteProof) and about epochKeyLiteHelper can be found [here](https://developer.unirep.io/docs/contracts-api/classes/typechain.factories.contracts.verifierHelpers.EpochKeyLiteVerifierHelper__factory)

## Tests

* [X] It should pass if success 
* [X] Should revert if proof is not valid ( include that if user does not sign up)
* [X] Should not reuse submitted proof (should revert)
* [X] Should not accept `epoch` != `current epoch` 

## Verification Steps

in root dir
1. `yarn install`
2. `yarn build`
3. `yarn contracts test`

## Todo

-   [X] Draft implementation for `leaveComment` & `editComment()`
-   [X] Unit tests for `leaveComment()`
-   [X] Unit tests for `editComment()`

## Checklist

-   [X] All new and existing tests pass
-   [X] I have updated the documentation (if applicable)
-   [X] My changes do not introduce new security vulnerabilities
-   [X] Run `yarn lint:fix`

### CODE REVIEW

File: packages/contracts/contracts/UnirepApp.sol [hunk 1]
```diff
@@ -2,7 +2,7 @@
 pragma solidity ^0.8.0;
 import {Unirep} from '@unirep/contracts/Unirep.sol';
 import {EpochKeyVerifierHelper} from '@unirep/contracts/verifierHelpers/EpochKeyVerifierHelper.sol';
-
+import {EpochKeyLiteVerifierHelper} from '@unirep/contracts/verifierHelpers/EpochKeyLiteVerifierHelper.sol';
```
Comment (NOOMA-42) [inline packages/contracts/contracts/UnirepApp.sol:5]: Could you briefly explain newly imported LiteVerifier in PR description and also link to doc for future reference?

File: packages/relay/src/synchornizer.ts [hunk 2]
```diff
@@ -42,6 +44,14 @@ export class UnirepSocialSynchronizer extends Synchronizer {
         }
     }
 
+    async resetDatabase() {
```
Comment (NOOMA-42) [inline packages/relay/src/synchornizer.ts:47]: Also add this new feat to your PR description?

File: packages/relay/src/index.ts [hunk 3]
```diff
@@ -3,7 +3,9 @@ import path from 'path'
 import fs from 'fs'
 import express from 'express'
 import { ethers } from 'ethers'
-import { SQLiteConnector } from 'anondb/node.js'
+import { SQLiteConnector, PostgresConnector } from 'anondb/node.js'
```
Comment (NOOMA-42) [inline packages/relay/src/index.ts:6]: why do we need 2 types of DB?

**Review by NOOMA-42 — APPROVED** [review]
LGTM in terms of code. I leave a few comment to ask for more detailed PR description

File: packages/contracts/contracts/UnirepApp.sol (hunk 1 above)
Comment (FoodChain1028) [reply packages/contracts/contracts/UnirepApp.sol:5]: Should I add a doc for smart contract in our repo? (I will add this as an issue, first create a empty file for now and add the document in the future.)

File: packages/relay/src/index.ts (hunk 3 above)
Comment (FoodChain1028) [reply packages/relay/src/index.ts:6]: umm, @NOOMA-42  this part is automatic merge from main, but I have no idea why there's this difference between the two file.

File: packages/contracts/contracts/UnirepApp.sol (hunk 1 above)
Comment (NOOMA-42) [reply packages/contracts/contracts/UnirepApp.sol:5]: > Should I add a doc for smart contract in our repo? (I will add this as an issue, first create a empty file for now and add the document in the future.)

IMO you can reference with just a link to unirep package.

File: packages/relay/src/index.ts (hunk 3 above)
Comment (Lisooo790926) [reply packages/relay/src/index.ts:6]: Yes this is for local and stage DB connection 
- Local => SQLiteConnector (memory DB)
- Stage => PostgresConnector (remote DB)

**Review by Lisooo790926 — APPROVED** [review]

**Pull request merged**